### ISSUE TITLE
Better control over Prometheus metrics

### ISSUE THREAD

**Issue by dtelyukh** [caddyserver/caddy#4016]
On one hand, with lack of lables, prometheus metrics make barely more sense than measuring nationwide average body temperatures. On the other hand, with too many labels, in prometheus might be a source of unbounded high-cardinality, which will cause issues when scraping with Prometheus over long periods of time and with larger deployments.

In the usecase we had, we needed to see per-host metrics (for an obvious reason, different hosts often have very different performance), and also a per-content-type metric, here's why:

1) Dynamically generated resources usually have way bigger TTFB, and mixing such resources with static files we obscure actual peaks in data, because each dynamically generated HTML document also pulls 10 to 100 static files. Content-type is a good heuristic for dynamic content, because html is usually dynamically generated, and everything else is static in most cases

2) No matter if html is dynamic or static, its performance is more critical for overall page load time, because it's the root document for the page, and each millisecond of its load time is always added to overall page load time, which is not always the case for page resources.

We intended to create two extra grafana panels, to be able to see metrics subsets that represent:
- html documents only
- small static files (js/css)
- large static files (images)

Also, it might be useful to separate metrics by document cache status, when caddy cache is enabled.

Based at the discussion in this ticket: https://github.com/caddyserver/caddy/issues/3784 and in this and in this PR: https://github.com/caddyserver/caddy/pull/4015

We came up with a following proposal:
- It's better to add extra configuration options, that allow both more detailed metrics and reduction of metrics size, rather just hardcoding extra added labels
- We cannot extend an existing `metrics` directive, because it defines the virtual host that serves metrics if defined inside of a specific host
- we chose `metrics_options` as new directive name; does anyone have better ideas maybe?

```
// enable per-host metrics for a single host
example.com {
  respond /foo "hello world"
  metrics_options {
    tags {
      host example.com
    }
  }
}
// enable per-host metrics + per-content-type for a single host
example2.com {
  respond /bar "hello world"
  metrics_options {
    tags {
      host example2.com
    }
    headers-to-tags {
      response.content-type
    }
  }
}
// disable metrics for a single host
example3.com:8443 {
  respond /baz "hello world"
  metrics_options {
    disable
  }
}
// disable metrics for a subpath
example3.com {
  handle_path /static {
    metrics_options {
      disable_metrics
    }
  }
}
// globally enable per-host metric, and also per-cache-status, based on response 
metrics_option {
  headers-to-tags {
    request.host
    response.x-cache-status
  }
}
```

**Comment by francislavoie** [comment]
I'm not sure that `metrics_options` can be per-path though (or even have matchers), because by that point it's too late. Metrics collection starts before the request is passed through the handlers, I think.

Also, please remember that what's important to think about is how it's configured in JSON, before even thinking of how it'll look in the Caddyfile. The Caddyfile is just an adapter to JSON config.

**Comment by dtelyukh** [comment]
> And I figure this would have to be a handler that defers execution until the middleware chain is coming back downstream if we plan to support reading from the response headers.
@francislavoie, it's OK, because Caddy already sends response code to metrics.

**Comment by hairyhenderson** [comment]
I haven't had time to think too deeply about this yet, but in general all labels should be set on all endpoints. That is to say, if a `content_type` or `host` label is tracked, it should be present for all time series.

Another way to say that is - no per-matcher/per-path metrics options, metric configuration should apply globally. The one potential exception to this is we may want to have some way to ignore certain endpoints. But that doesn't need to be part of this design.

I don't think it's necessary to have a new directive - we should be able to use the existing `metrics` directive.

And, I like the idea that @francislavoie implied with having custom labels that simply use a placeholder (or even a static value) as the value; I could see a directive like this:

```
metrics /metrics {
	label host {http.request.host}
	label content_type {http.response.header.Content-Type}
	label static_label "same value for everyone!"
}
```

I'm not at all familiar with the placeholder code but I _think_ this would be relatively straightforward to do. I'll see if I can find some time in the next few weeks to work on this, but I'd be very open to contributions for a design like this 😅

**Comment by francislavoie** [comment]
> Another way to say that is - no per-matcher/per-path metrics options, metric configuration should apply globally.

If that's the case, then the existing `metrics` handler won't do, because in the Caddyfile, a site block _is_ a `host` matcher. Metrics would need to be configured via global options, and probably be set at the `apps.http.servers` level. (FWIW if it's a server option, it can be made part of https://caddyserver.com/docs/caddyfile/options#server-options)

**Comment by francislavoie** [comment]
I spoke with @hairyhenderson and ultimately I don't think this'll be possible in any way that doesn't have drastic downsides.

The trouble is that metrics are configured at process startup, and can't be safely reconfigured because the prometheus metrics lib aims to remain consistent for the lifetime of the process. If a reconfiguration were to happen, then the entire metrics storage would need to be thrown away and reinitialized. This is not ok, because certain counters need to remain because they track other global state.

Since Caddy is designed to have reloadable config, without restarting the process, this is at odds with that idea. Theoretically we could delegate initialization of the metrics until after the config is initially loaded, but this would mean that any subsequent config change that attempts to change metrics config would have to be rejected. Not ideal. Breaks certain assumptions made in [Caddy's architecture](https://caddyserver.com/docs/architecture).

So to avoid needing to open the possibility for reloads to fail, any configuration for metrics would have to be provided separately, specifically at process startup. This might mean having a separate config mechanism outside of the regular configs (e.g. environment variables, or new CLI args to `caddy run`), but this has pretty bad UX problems.

So... what do you think @mholt? I don't see a happy path here.

**Comment by codecat** [comment]
> The trouble is that metrics are configured at process startup, and can't be safely reconfigured because the prometheus metrics lib aims to remain consistent for the lifetime of the process.

I'm not at all familiar with the metrics lib used, but would it be possible to contribute a patch to the lib that allows re-configuration at runtime?

**Comment by francislavoie** [comment]
I don't think it would be easy. It seems fundamental to how its designed. The lib in question is https://github.com/prometheus/client_golang/tree/master/prometheus.

I know that @hairyhenderson or I aren't interested in working on that, so someone else would need to work on it if they're up for it.

**Comment by hairyhenderson** [comment]
Just to add a bit more clarity on the issue - when exposing metrics with Prometheus, the expectation is that all metrics will remain consistent for the lifetime of the process, both in terms of value (a counter which starts at `0` at the beginning of the process should continue to increase indefinitely, as a reset back to `0` is interpreted as a process restart), and also in terms of labels (if `caddy_http_request_duration_seconds_bucket{code="308",handler="static_response",method="GET",server="remaining_auto_https_redirects",le="0.005"}` is exposed, it must not gain or lose a label during the lifetime of the process).

Another thing to consider is that Caddy exposes metrics on the admin API. An important metric for monitoring Caddy is the `caddy_admin_http_request_errors_total` metric. When there's an error during config reload, the `caddy_admin_http_request_errors_total{handler="load",method="POST",path="/load"}`series will increase, and it would be reasonable for a Prometheus alert to be set to watch that series and alert if the count of errors goes up above a desired threshold. Resetting this counter risks losing any visibility on errors.

Bottom line: we can't move forward with any customization of the metrics series (i.e. labels) until we can find a suitable config mechanism that doesn't require resetting at config reload time. I hope @mholt has some ideas on that.

**Comment by mholt** [comment]
@hairyhenderson

> we can't move forward with any customization of the metrics series (i.e. labels) until we can find a suitable config mechanism that doesn't require resetting at config reload time.

That shouldn't be a problem. Caddy has something called a [`UsagePool`](https://pkg.go.dev/github.com/caddyserver/caddy/v2#UsagePool) that is designed for the sole purpose of maintaining global state across config reloads. You can dump your stats into that. This is how Caddy keeps track of the health status of proxy backends between config reloads, for example.

**Comment by francislavoie** [comment]
@mholt I think you misunderstood, metrics are _already_ global. The problem is that the metrics _cannot ever change_ after the first load.

**Comment by mholt** [comment]
@francislavoie What does their scope have to do with whether they change or not?

**Comment by francislavoie** [comment]
Because the only way to change scope involves resetting the storage. As we're written above, that's a deal-breaker, we can't reset the counters during config reloads.

**Comment by mholt** [comment]
That sounds like a pretty severe limitation in Prometheus, which is surprising. I don't know how I could be of help here.

**Comment by francislavoie** [comment]
It's not a "bug" though, it's an assumption that underlies the metrics tracking system.

The question is this:

> So to avoid needing to open the possibility for reloads to fail, any configuration for metrics would have to be provided separately, specifically at process startup. This might mean having a separate config mechanism outside of the regular configs (e.g. environment variables, or new CLI args to caddy run), but this has pretty bad UX problems.

and

> Bottom line: we can't move forward with any customization of the metrics series (i.e. labels) until we can find a suitable config mechanism that doesn't require resetting at config reload time. I hope @mholt has some ideas on that.

**Comment by mholt** [comment]
It's a broken assumption then. Servers change their configs, it's as simple as that. Prometheus should know that by now.

**Comment by mr-karan** [comment]
We can probably use https://github.com/VictoriaMetrics/metrics which supports creating metrics on the fly. I believe what you want to use is creating a [`NewSet`](https://github.com/VictoriaMetrics/metrics/blob/master/set.go#L24).

It's also a nicer, less dependency alternative to the official client lib.

cc @hairyhenderson @francislavoie

**Comment by hairyhenderson** [comment]
@mr-karan I'm not sure that's exactly what we need. What's lacking is not the ability to create metrics "on the fly" (which the Prometheus client can obviously also do), but instead the ability to _modify_ metrics after being created.

If that's possible with VictoriaMetrics, then that indicates a bug in that client, not a feature. Metrics need to be guaranteed to stay consistent through the life of the process.

**Comment by mr-karan** [comment]
> but instead the ability to modify metrics after being created

To demonstrate with an example:

`metric_label_a{val="ok"} 1`

You want this to be:

`metric_label_a{new_val="hello"} 1`

Is my understanding correct?

**Comment by Manoj321** [comment]
Leaving this here in case it helps - If you use telegraf to scrape metrics you can easily rename the tag values: https://github.com/Manoj321/ansible-synology/commit/83ca673fcd1f949dbb9f20dd2b3fc7cddd4ef8ef

**Comment by danlsgiga** [comment]
This issue doesn't seem to be getting traction. It would be nice to have the labels capability.
In my case, I have the config via Caddyfile and metrics are emitted using the listener server name (`srv0`, `srv1`, etc) as a label. Since I can't change the listener server name via Caddyfile, only JSON, its hard for me to know what metrics relate to what listeners.

**Comment by mediocregopher** [comment]
Just chiming in a few years late, I read through this thread I and I'm not sure that I agree with the assumption that prometheus metrics absolutely cannot have their labels modified or their counters reset to 0 once a process has started. If the metrics registry is completely reset on reload, from prometheus's perspective that will be indistinguishable from a restart, and of course prometheus deals with processes restarting just fine. What would be required is to create a new `Registry` object during provisioning, use that when provisioning all modules during the reload, and at the end of the reload swap the `Registry` that's being used to serve `/metrics` to use the new one.

-----

Ignoring that issue, I went ahead and implemented a pretty full-featured metric plugin as part of my plugin suite:

https://dev.mediocregopher.com/mediocre-caddy-plugins/#httphandlersrequest_timing_metric-response_size_metric

It supports custom labels, including using placeholders, response matching, and custom buckets. The new metrics get added into caddy's builtin `Registry` that's found on the `caddy.Context`. The configuration of the histograms is disconnected from the middleware handlers which populate them, so it's fairly flexible in what you can accomplish.

I'm posting this just to show an example of how this sort of thing might be implemented, though I'm sure my own implementation is a bit rough and not fully idiomatic with the rest of Caddy's codebase.

**Comment by roganartu** [comment]
I think the information in this issue is a little outdated. The above comment is correct that prometheus handles metrics resetting just fine (though regular resetting can cause some issues with accuracy depending on how frequently resets occur and how wide a particular prometheus query time window is), but it's also possible to just outright delete metrics by unique label sets.

Assuming the following:
- Caddy can keep track of the set of labels for each metric from the previous iteration of the config
- The set of labels for a metric that might need to be removed during a reload is knowable (ie: not dynamic based on request information, but static based on values from config like host, site, handler etc)
- Caddy can compare that set of current labels with the set of labels from the reloaded config

then https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#MetricVec.Delete or similar can be used on each metric to remove the labels that should no longer be exported. This will remove those specific observations without resetting all counters.
